### ISSUE TITLE
kubernetes openstack cloud provider seems have hardcode network name

### ISSUE THREAD

**Issue by HackToday** [kubernetes/kubernetes#18409]
From https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/openstack/openstack.go 
we can find that,
```golang
func findAddrs(netblob interface{}) []string {
	// Run-time types for the win :(
	ret := []string{}
	list, ok := netblob.([]interface{})
	if !ok {
		return ret
	}
	for _, item := range list {
		props, ok := item.(map[string]interface{})
		if !ok {
			continue
		}
		tmp, ok := props["addr"]
		if !ok {
			continue
		}
		addr, ok := tmp.(string)
		if !ok {
			continue
		}
		ret = append(ret, addr)
	}
	return ret
}
```

1. the hardcode "private" and "public", seems only support that network with such name.

2. second, it fetch the first ip. usually, it can include both v6 and v4 ip. what's the meaning to just fetch the first one ? is it v4 or v6 ?

**Comment by HackToday** [comment]
ping @anguslees and @jsafrane

**Comment by anguslees** [comment]
Yes.  Zoom out a level or two in the code, and you'll see we're trying to either a) setup an openstack loadbalancer to point to the minion node (a single address) using `getAddressByName()`, or b) report to kubernetes all the addresses we know about the node and whether they are "internal" or "external" in `NodeAddresses()`.  All we have at the starting point is the server name (from kubernetes), and we need to make up the rest of this information somehow.

(I agree (a) should reuse the NodeAddresses() function from (b), but the latter is a newer function and the code hasn't yet been refactored in that way.  This doesn't really change the substance of this discussion.)

Re (1):

The logic as currently implemented is:

- Assume any network named "private" is "internal"
- Assume any network named "public" is "external"
- Assume the AccessIPs (IPv4 and IPv6) are "external"
- Any other network addresses are ignored.

I implemented the first two ("private" and "public") following the Rackspace public cloud precedent, and  on the basis that users in other OpenStack clouds can use these names if they want to encourage an address other than the Access addresses.

For the (a) use case, we just take the first address we find in order of the above; for (b) we report that full list.  I preferred "private" to "public" since (again, at least on Rackspace public cloud), the traffic on "private" network is sometimes cheaper/free.

I agree this isn't great, and please suggest something better if you have a use case in mind.  Perhaps we should pull a list of network names that should be considered "internal" and "external" from the provider config?  Perhaps I could assume that provider networks are public and non-provider networks are internal?  I don't want to add too much configuration here if there's a good algorithm that will Just Work.

I think in practice it's probably ok to continue to prefer "internal" to "external" for loadbalancer setup, but I can contrive hypothetical cases that need more flexibility here too.  Again, if you have a concrete need for such a thing, please file a bug so we have some idea of the flexibility required.

Complicating all this is that Kubernetes never actually defined what an "internal" address _was_.  `docs/admin/node.md` just says "Generally the IP address of the node that is routable only within the cluster" which isn't a sufficiently strict definition.  Looking at where Kubernetes actually uses it:
- `GetNodeHostIP()` just uses the first address reported (preferring InternalIPs to ExternalIPs) and
- the Kubernetes master configures an "SSH Tunnel Check" which just uses the first ExternalIP

... which kind of makes it pointless to ever even bother reporting anything beyond the first address.  In both cases, it's assumed that this address is reachable by the Kubernetes master nodes, which makes the Internal vs External distinction a bit pointless.

Re (2):

`NodeAddresses()` is meant to return more than one IP, and the loadbalancer setup can only use a single IP for each node (or would unnecessarily track/healthceck multiple paths to the same backend for multiple addresses).  If you're seeing single addresses returned by `NodeAddresses()` then you should file a bug.

All the above treats IPv4 and IPv6 as equally important, if the OpenStack cloud has both, and the "first" address used for the loadbalancer will be an arbitrary choice between the two protocols (not quite true: `getAddressByName()` will use AccessIPv4 before AccessIPv6).  I think that's how it _should_ be (the rest of the code should be protocol agnostic), but I wouldn't be surprised if there's some part of Kubernetes that doesn't support IPv6 properly.  There certainly used to be _lots_ of places that assumed IPv4 in Kubernetes, and GCE doesn't support IPv6 iirc so it gets less testing - if you've found something that breaks with IPv6, _please_ file bugs.

---

Now this bug: what would you like done here?  Would you like the current situation to be documented further? Is the "private"/"public" naming thing too restrictive for your usecase? Something else?

**Comment by HackToday** [comment]
Hi, if we talked about openstack, I expect it not bind to rackspace, although, the kubernetes called RK SDK,  But from my understanding, seems RK itself SDK not bind network to "private" and "pubic". User can define public network with their names, like site-network, for example. or "testnet", 
And for private, it should be seem.


For nova list, the addresses, it can have  "OS-EXT-IPS:type":"fixed" or  "OS-EXT-IPS:type":"floating"
for floating, I think what you means "Public", for fixed, I think its what you means "Private".

In other words, we'd better to find related information from other fields in nova or neutron infromation. Instead of bind to "private" and "public" name.

**Comment by lebauce** [comment]
I faced this issue and wrote a patch that implements what @HackToday suggests. I will send a PR soon

**Comment by HackToday** [comment]
Glad to see some improvements here. And wish to use it if implemented
